lib/smack/RewriteUndefPtrs.cpp

```diff
@@ -0,0 +1,41 @@
+//===----------------------------------------------------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+#include "smack/RewriteUndefPtrs.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+namespace smack {
+
+char RewriteUndefPtrs::ID = 0;
+static RegisterPass<RewriteUndefPtrs>
+    X("rewrite-undef-ptrs", "Transform undef pointers into null pointers");
+
+bool RewriteUndefPtrs::runOnFunction(Function &F) {
+  bool changed = false;
+  for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+    for (auto OI = I->op_begin(), OE = I->op_end(); OI != OE; ++OI) {
```

Comment:
Is it possible that undefined pointers are not direct operands? Namely, they can show up in some expressions which are the operands of instructions. Maybe we need to do a BFS/DFS here.

Reply:
This is possible. For example if we have `%1 = { i8*, i8* } undef` then extracted pointers are `undef`. This is why it appears after the SplitAggregateValue pass. A better solution might be to turn `undef` constants into calls to an undef initializer like what the sea-dsa devs mentioned.